File: eclipse-plugins/edu.wpi.first.wpilib.plugins.cpp/resources/templates/examples/GearsBot/src/Subsystems/Wrist.cpp
#include "Wrist.h"
#include "SmartDashboard/SmartDashboard.h"
#include "LiveWindow/LiveWindow.h"

Wrist::Wrist() : PIDSubsystem("Wrist", kP_real, 0.0, 0.0) {
	#ifdef SIMULATION // Check for simulation and update PID values
        GetPIDController()->SetPID(kP_simulation, 0, 0, 0);
	#endif
    SetAbsoluteTolerance(2.5);

    motor = new Victor(6);

    // Conversion value of potentiometer varies between the real world and simulation
	#ifdef REAL
        pot = new AnalogPotentiometer(3, -270.0/5);
	#else
        pot = new AnalogPotentiometer(3); // Defaults to degrees
	#endif

	// Let's show everything on the LiveWindow
    // TODO: LiveWindow::GetInstance()->AddActuator("Wrist", "Motor", (Victor) motor);
    // TODO: LiveWindow::GetInstance()->AddSensor("Wrist", "Pot", (AnalogPotentiometer) pot);
    LiveWindow::GetInstance()->AddActuator("Wrist", "PID", GetPIDController());
}

void Wrist::Log() {
    // TODO: SmartDashboard::PutData("Wrist Angle", (AnalogPotentiometer) pot);
}

double Wrist::ReturnPIDInput() {
    return pot->Get();
}

void Wrist::UsePIDOutput(double d) {
    motor->Set(d);
}

File: wpilibj-examples/src/main/java/edu/wpi/first/wpilibj/examples/gearsbot/subsystems/Wrist.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) FIRST 2017. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj.examples.gearsbot.subsystems;

import edu.wpi.first.wpilibj.AnalogPotentiometer;
import edu.wpi.first.wpilibj.SpeedController;
import edu.wpi.first.wpilibj.Victor;
import edu.wpi.first.wpilibj.command.PIDSubsystem;
import edu.wpi.first.wpilibj.examples.gearsbot.Robot;
import edu.wpi.first.wpilibj.interfaces.Potentiometer;
import edu.wpi.first.wpilibj.livewindow.LiveWindow;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 * The wrist subsystem is like the elevator, but with a rotational joint instead
 * of a linear joint.
 */
public class Wrist extends PIDSubsystem {

	private SpeedController m_motor;
	private Potentiometer m_pot;

	private static final double kP_real = 1;
	private static final double kP_simulation = 0.05;

	public Wrist() {
		super(kP_real, 0, 0);
		if (Robot.isSimulation()) { // Check for simulation and update PID values
			getPIDController().setPID(kP_simulation, 0, 0, 0);
		}
		setAbsoluteTolerance(2.5);

		m_motor = new Victor(6);

		// Conversion value of potentiometer varies between the real world and
		// simulation
		if (Robot.isReal()) {
			m_pot = new AnalogPotentiometer(3, -270.0 / 5);
		} else {
			m_pot = new AnalogPotentiometer(3); // Defaults to degrees
		}

		// Let's show everything on the LiveWindow
		LiveWindow.addActuator("Wrist", "Motor", (Victor) m_motor);
		LiveWindow.addSensor("Wrist", "Pot", (AnalogPotentiometer) m_pot);
		LiveWindow.addActuator("Wrist", "PID", getPIDController());
	}

	@Override
	public void initDefaultCommand() {
	}

	/**
	 * The log method puts interesting information to the SmartDashboard.
	 */
	public void log() {
		SmartDashboard.putData("Wrist Angle", (AnalogPotentiometer) m_pot);
	}

	/**
	 * Use the potentiometer as the PID sensor. This method is automatically
	 * called by the subsystem.
	 */
	@Override
	protected double returnPIDInput() {
		return m_pot.get();
	}

	/**
	 * Use the motor as the PID output. This method is automatically called by
	 * the subsystem.
	 */
	@Override
	protected void usePIDOutput(double power) {
		m_motor.set(power);
	}
}
